### PR TITLE
[SPARK-37332][SQL] Allow ANSI intervals in `ALTER TABLE .. ADD COLUMNS`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to allow ANSI intervals: year-month and day-time intervals in the `ALTER TABLE .. ADD COLUMNS` command for tables in v1 and v2 In-Memory catalogs. Also added an unified test suite to migrate related tests in the future.

### Why are the changes needed?
To improve user experience with Spark SQL. After the changes, users will be able to add columns with ANSI intervals instead of dropping and creating new table.

### Does this PR introduce _any_ user-facing change?
In some sense, yes. After the changes, the command doesn't output any error message.

### How was this patch tested?
By running new test suite:
```
$ build/sbt -Phive-2.3 -Phive-thriftserver "test:testOnly *AlterTableAddColumnsSuite"
$ build/sbt -Phive-2.3 "test:testOnly *HiveDDLSuite"
```

### CODE REVIEW

**Comment by sarutak** [comment]
> v1 Hive external catalog doesn't support the command

v1 Hive external catalog supports `ALTER TABLE ADD COLUMN` doesn't it?

```
spark-sql> CREATE TABLE tbl1(x INT) using Parquet;
spark-sql> ALTER TABLE tbl1 ADD COLUMN (y STRING);
spark-sql> DESCRIBE tbl1;
x                   	int                 	                    
y                   	string  
```

**Comment by MaxGekk** [comment]
> v1 Hive external catalog supports ALTER TABLE ADD COLUMN doesn't it?

Thanks. I have corrected the description.

**Comment by MaxGekk** [comment]
@cloud-fan @sarutak @Peng-Lei @AngersZhuuuu Could you review this PR, please.

**Comment by MaxGekk** [comment]
@cloud-fan FYI, Here is the ticket for fully unified tests of `ALTER TABLE .. ADD COLUMNS` https://issues.apache.org/jira/browse/SPARK-37045

**Comment by sarutak** [comment]
> v1 Hive external catalog doesn't support ANSI intervals

I think Hive external catalog supports ANSI intervals. In this case, the format of metadata is not Hive compatible but Spark specific.
```
spark-sql> CREATE TABLE tbl2(ym INTERVAL YEAR TO MONTH, dt INTERVAL DAY TO SECOND) using Parquet;
21/11/15 21:05:08 WARN HiveExternalCatalog: Hive incompatible types found: interval year to month, interval day to second. Persisting data source table `default`.`tbl2` into Hive metastore in Spark SQL specific format, which is NOT compatible with Hive.
spark-sql> INSERT INTO tbl2 VALUES(INTERVAL '1-2' YEAR TO MONTH, INTERVAL '1 2:3:4' DAY TO SECOND);
spark-sql> SELECT * FROM tbl2;
1-2	1 02:03:04.000000000
Time taken: 0.381 seconds, Fetched 1 row(s)

```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddColumnsSuiteBase.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * This base suite contains unified tests for the `ALTER TABLE .. ADD COLUMNS` command that
+ * check V1 and V2 table catalogs. The tests that cannot run for all supported catalogs are
+ * located in more specific test suites:
+ *
+ *   - V2 table catalog tests:
+ *     `org.apache.spark.sql.execution.command.v2.AlterTableAddColumnsSuite`
+ *   - V1 table catalog tests:
+ *     `org.apache.spark.sql.execution.command.v1.AlterTableAddColumnsSuiteBase`
+ *     - V1 In-Memory catalog:
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddColumnsSuiteBase.scala:33]: We do not have one suite for hive catalog?

**Comment by MaxGekk** [comment]
> We do not have one suite for hive catalog?

@cloud-fan I have added one.

**Review by sarutak — APPROVED** [review]
LGTM. Pending CIs.

**Comment by sarutak** [comment]
I cloned `MaxGekk:add-columns-ansi-interval` and run GA jobs on my repository, and `docker integration tests` passed.
https://github.com/sarutak/spark/runs/4219609621?check_suite_focus=true

So, I think the `docker integration tests` failure on your repository is not related to this change.

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @sarutak and @cloud-fan for review.

**Pull request closed without merging**